### PR TITLE
Archive Block: Prevent spacing styles and additional CSS classes from being printed twice

### PR DESCRIPTION
**Pull request by t-hamano**
Follow-up on: 

- #43935
- #43350

Similar to #44439
Blocker: #44491

_Note_: I think the double spacing is a major layout issue, so it would be a good idea to be backported to WordPress 6.1.

_@edit_: As per the discussion that began with [this comment](https://github.com/WordPress/gutenberg/pull/44438#pullrequestreview-1119477064), it was decided to move the logic for filtering the block support attributes to the `ServerSideRender` component itself. Once the appropriate prop is added to the `ServerSideReder` component, the code changes in this PR will be simpler. This PR will be placed on freeze until a PR regarding improvements to `ServerSideRender` is submitted and merged.

_@update_: Now that #44491 has been merged, the new `skipBlockSupportAttributes ` prop of the `ServerSideRender` component is available. This allows us to solve the problem this PR addresses in a simpler way.

## What?
This PR fixes a problem in the editor where block support inline styles and additional CSS classes are rendered twice.

## Why?
This is because this block uses the `ServiceSideRender` component and receives all attributes (including styles generated by the block support).

## How?
As the discussion started with [this comment](https://github.com/WordPress/gutenberg/issues/43243#issuecomment-1229114714), there are many possible approaches for a block using the `ServiceSideRender` component. I have tried to solve this by not passing block support styles and additional classes to the `ServiceSideRender` component.

## Testing Instructions
- Add an archive block.
- Apply some block support styles (Typography, Dimensions).
- Add additional CSS class.
- Confirm that the block support styles and additional CSS classes aren't applied to the `ServiceSideRender` component and are rendered only in the block wrapper by using the browser console.

### Before

before

### After

after

### CODE REVIEW

File: packages/block-library/src/archives/edit.js [hunk 1]
```diff
@@ -14,6 +14,17 @@ import ServerSideRender from '@wordpress/server-side-render';
...
 	const { showLabel, showPostCounts, displayAsDropdown, type } = attributes;
 
+	// Prevent block support styles and additional CSS classes from being printed twice.
+	const serverSideRenderAttirbutes = {
+		...attributes,
+		className: undefined,
+		style: {
+			...attributes.style,
+			spacing: undefined,
+			typography: undefined,
```
Comment (t-hamano) [inline packages/block-library/src/archives/edit.js]: If border support is added in the future, it will be necessary to write `border: undefined` here.

File: packages/block-library/src/archives/style.scss [hunk 2]
```diff
@@ -1,3 +1,8 @@
+.wp-block-archives {
+	// This block has customizable padding, border-box makes that more predictable.
+	box-sizing: border-box;
```
Comment (t-hamano) [inline packages/block-library/src/archives/style.scss:3]: Added according to #43465, but will revert if this fix is not needed now.

File: packages/block-library/src/archives/edit.js [hunk 3]
```diff
@@ -14,6 +14,17 @@ import ServerSideRender from '@wordpress/server-side-render';
 export default function ArchivesEdit( { attributes, setAttributes } ) {
 	const { showLabel, showPostCounts, displayAsDropdown, type } = attributes;
 
+	// Prevent block support styles and additional CSS classes from being printed twice.
+	const serverSideRenderAttirbutes = {
```
Comment (aaronrobertshaw) [inline packages/block-library/src/archives/edit.js]:
```suggestion
	const serverSideRenderAttributes = {
```

Typo.

File: packages/block-library/src/archives/edit.js [hunk 4]
```diff
@@ -66,7 +77,7 @@ export default function ArchivesEdit( { attributes, setAttributes } ) {
 				<Disabled>
 					<ServerSideRender
 						block="core/archives"
-						attributes={ attributes }
+						attributes={ serverSideRenderAttirbutes }
```
Comment (aaronrobertshaw) [inline packages/block-library/src/archives/edit.js]:
```suggestion
						attributes={ serverSideRenderAttributes }
```

Typo.

File: packages/block-library/src/archives/edit.js (hunk 1 above)
Comment (aaronrobertshaw) [reply packages/block-library/src/archives/edit.js]: This doesn't filter out the `fontSize` or `fontFamily` attributes that are also used by the typography supports to generate CSS classnames. The result is that classes for those are still duplicated which goes against the stated intent of this PR.

As there are a few top-level attributes added by block supports and any block using `ServerSideRender` needs to remove them along with the style attribute, it seems like we might be able to address this [within `ServerSideRender` itself](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/packages/server-side-render/src/server-side-render.js#L91) rather than duplicate that logic across several blocks.

As a quick guide the following top-level attributes added by block supports come to mind;
- `fontSize`
- `fontFamily`
- `borderColor`
- `backgroundColor`
- `textColor`
- `gradient`

There are probably some others as well for things like align support etc.

**Review by aaronrobertshaw — COMMENTED** [review]
Thank you for putting this PR together @t-hamano 👍 

I ran into one issue while testing this. It doesn't remove the duplicated CSS classes as advertised. For typography classes, it would need to remove `fontSize` and `fontFamily` top-level attributes.

Screen Shot 2022-09-26 at 10 21 42 am

Taking a step back though, I do wonder if we'd be better served to move this logic of filtering out the block support attributes to the `ServerSideRender` component itself. It saves us from having to duplicate this same logic across several blocks. What do you think?

I'd also like to touch base with @kevin940726 and @andrewserong, who have either worked on `ServerSideRender` or looked at issues on this front previously, to get their thoughts.

**Comment by andrewserong** [comment]
> Taking a step back though, I do wonder if we'd be better served to move this logic of filtering out the block support attributes to the ServerSideRender component itself. It saves us from having to duplicate this same logic across several blocks. What do you think?

Thanks for the ping! From a quick read of the `ServerSideRender` readme, it looks like it's encouraging us not to use it:

> ServerSideRender should be regarded as a fallback or legacy mechanism, it is not appropriate for developing new features against. New blocks should be built in conjunction with any necessary REST API endpoints...

It also looks like that package hasn't been updated much, so it seems the general idea from other contributors was that we should be updating blocks to not rely on the `ServerSideRender` component and instead be refactoring to a real-time JS-based edit view of the block, when we're adding features (like block supports)?

So I suppose I'd think of the fix in this PR (or an update in `ServerSideRender`) as being a shorter-term fix, rather than the ideal end state. But I think you make a good point about updating `ServerSideRender` to do the skipping. If there's a consistent structure we want across each of the blocks that currently use the component, perhaps we could add the logic to `ServerSideRender` and then a boolean prop like `skipBlockSupportStyles` to switch on the skipping?

From a quick skim, the blocks that currently use `ServerSideRender` are:

* Archives
* Calendar
* Latest Comments
* RSS
* Tag Cloud

For some of these, I wonder how hard it'd be to do a JS-based version of the server-rendered markup? 🤔 I imagine something like Tag Cloud is likely not really worth the effort to duplicate the PHP, so now that I've written out this comment, I think I know why we're currently using `ServerSideRender` 😅

TL;DR: Yes, I think the idea of consolidating the logic of which attributes to skip for block supports styles is a good idea 👍

**Comment by aaronrobertshaw** [comment]
Thanks for the extra input @andrewserong 🙇 

> From a quick read of the ServerSideRender readme, it looks like it's encouraging us not to use it:

My view is the blocks you noted are all old existing blocks that began using `ServerSideRender` about 2 years ago. Prior to a lot of the block supports that are in play today. I think it makes sense to update `ServerSideRender` as discussed with the long term view, as you say, to refactor these blocks away from `ServerSideRender` and towards a real-time JS based version.

Also, if we've released the `@wordpress/server-side-render` package, it would make sense that it could be used out in the wild despite encouragement against its use. In those cases, they'd also likely be getting the extra block support styles and classes.

**Comment by t-hamano** [comment]
Thank you for the advice, @aaronrobertshaw, @andrewserong !

> My view is the blocks you noted are all old existing blocks that began using ServerSideRender about 2 years ago. Prior to a lot of the block supports that are in play today. I think it makes sense to update ServerSideRender as discussed with the long term view, as you say, to refactor these blocks away from ServerSideRender and towards a real-time JS based version.

I also agree with this opinion.

According to [WPdirectory's search results](https://wpdirectory.net/search/01GDWH8TKEJ71F28VH5J9NYXRV), there are several plugins that use `ServerSideRender`. Adding a property like `skipBlockSupportStyles` to `ServerSideRender` would make sense, as it would also provide a way for the developer to smoothly opt-in to block support in the future.

As for the core blocks, I think they should be refactored to be as JS-based as possible, but this is a long-term thing, and for now the priority should be to keep `ServerSideRender` and render global styles properly.

If you are comfortable with this direction, I would freeze #44438 and #44439 and consider a new PR to extend ServerSideRender, what do you think? 🙂

**Comment by andrewserong** [comment]
> If you are comfortable with this direction, I would freeze https://github.com/WordPress/gutenberg/pull/44438 and https://github.com/WordPress/gutenberg/pull/44439 and consider a new PR to extend ServerSideRender, what do you think? 🙂

Sounds good to me, thanks for your flexibility here @t-hamano!

**Comment by aaronrobertshaw** [comment]
> If you are comfortable with this direction, I would freeze https://github.com/WordPress/gutenberg/pull/44438 and https://github.com/WordPress/gutenberg/pull/44439 and consider a new PR to extend ServerSideRender, what do you think? 🙂

Sounds like a plan to me also. I'll happily review the new PR when it's ready 👍

**Comment by t-hamano** [comment]
I have added the new `skipBlockSupportAttributes` prop introduced by #44491.
As per the testing procedure described in the first comment of this PR, this issue should be resolved.

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for cleaning this up.

It tests well for me. I could confirm the duplicate classes and styles before checking out this PR. After doing so, the duplicates were omitted.

LGTM 🚢
